### PR TITLE
fix: Use splat syntax for cluster name to avoid `(known after apply)` in managed node groups

### PR DESCRIPTION
**Pull request by barryib**
# PR o'clock

## Description

After some tests, I've noticed that all inputs of the `null_data_source` are already know.

```shell
  # module.eks.data.null_data_source.node_groups[0] will be read during apply
  # (config refers to values not yet known)
 <= data "null_data_source" "node_groups"  {
      + has_computed_default = (known after apply)
      + id                   = (known after apply)
      + inputs               = {
          + "aws_auth"        = "kube-system/aws-auth"
          + "cluster_name"    = "test-eks-uIVr3XSp"
          + "role_CNI_Policy" = "test-eks-uIVr3XSp20200505210731845200000005-20200505210733402400000008"
          + "role_Container"  = "test-eks-uIVr3XSp20200505210731845200000005-20200505210733399800000007"
          + "role_NodePolicy" = "test-eks-uIVr3XSp20200505210731845200000005-20200505210733441800000009"
        }
      + outputs              = (known after apply)
      + random               = (known after apply)
    }
```
The unknown values are computed one like `random`, `id`, etc. and those will be read only during apply unless we remove the `count`. In that way Terraform will always read datas before apply and avoid the `(known after apply)` issue. Sounds like a Terraform magic again 🤷‍♂️

Resolves https://github.com/terraform-aws-modules/terraform-aws-eks/issues/843 #862 

### Checklist

- [x] CI tests are passing
- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by barryib** [comment]
> It works. And fixes the issues 👍
> 
> I didn't know that TF 0.12 supported depends_on for variables. I thought this hadn't been implemented yet and the bug traffic I found had it potentially scheduled for 0.13. I would have gone that route if I'd known it existed when first implementing this module 😉

So should we go through https://github.com/terraform-aws-modules/terraform-aws-eks/pull/867 or continue on this PR ?

File: node_groups.tf [hunk 1]
```diff
@@ -1,10 +1,8 @@
 # Hack to ensure ordering of resource creation. Do not create node_groups
 # before other resources are ready. Removes race conditions
 data "null_data_source" "node_groups" {
-  count = var.create_eks ? 1 : 0
-
   inputs = {
-    cluster_name = aws_eks_cluster.this[0].name
+    cluster_name = coalescelist(aws_eks_cluster.this[*].name, [""])[0]
```
[the remark this answers is not in the document]
Comment (barryib) [reply node_groups.tf:5]: So are you saying if we put back the `count = var.create_eks ? 1 : 0`, and just use `coalescelist(aws_eks_cluster.this[*].name, [""])[0]` it'll work ? It'll only update and no recreate ?

File: node_groups.tf [hunk 2]
```diff
@@ -1,10 +1,8 @@
 # Hack to ensure ordering of resource creation. Do not create node_groups
 # before other resources are ready. Removes race conditions
 data "null_data_source" "node_groups" {
-  count = var.create_eks ? 1 : 0
```
[the remark this answers is not in the document]
Comment (barryib) [reply node_groups.tf:4]: You're right.

File: node_groups.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (barryib) [reply node_groups.tf:7]: I also don't understand it.

**Comment by barryib** [comment]
> > It works. And fixes the issues 👍
> > I didn't know that TF 0.12 supported depends_on for variables. I thought this hadn't been implemented yet and the bug traffic I found had it potentially scheduled for 0.13. I would have gone that route if I'd known it existed when first implementing this module 😉
> 
> @dpiddockcmp So should we go with #867 or continue on this PR ?

Let's stick with this. It sounds to work without notable change.

**Comment by babilen5** [comment]
Just wanted to mention that I find the approach in #867 to be much more explicit and easier to understand. While it is true that the PR consists of more notable changes to the make-up of the module, it is more explicit, easier to understand and less "magic" than the approach here.

I have tested the changes in here and can confirm that they address the issues at hand, but I am unsure as to why that is exactly.

**Pull request merged**